sphinx/domains/std.py

```diff
@@ -223,12 +228,22 @@ def add_target_and_index(self, firstname: str, sig: str, signode: desc_signature
             node_id = make_id(self.env, self.state.document, prefix, optname)
             signode['ids'].append(node_id)
 
+            if 'compat-id' in self.options:
+                old_node_id = self.make_old_id(prefix, optname)
+                if old_node_id not in self.state.document.ids and \
+                   old_node_id not in signode['ids']:
+                    signode['ids'].append(old_node_id)
+
         self.state.document.note_explicit_target(signode)
 
         domain = cast(StandardDomain, self.env.get_domain('std'))
         for optname in signode.get('allnames', []):
-            domain.add_program_option(currprogram, optname,
-                                      self.env.docname, signode['ids'][0])
+            if 'compat-id' in self.options:
```

Comment:
I think it's enough to append an old-styled ID to the signode to keep old hyperlinks available. You don't need to register an old-styled ID to the domain.

Reply:
Done! I totally missed that for every `signode` added the html writer generates a span with the proper `id`, so removed the setting.